Mathlib/Data/Array/extractLemmas.lean

```diff
@@ -0,0 +1,142 @@
+  intro i s e h
+  rw [Array.size]
+  apply Array.extract_len_aux _ _ _ h
+
+theorem extract_loop_0 {array: Array α }: Array.extract.loop array 0 i e = e := by
+  unfold Array.extract.loop
+  simp
+
+theorem extract_end_le_start_eq_empty {a: Array α } :
+    (j ≤ i) → a.extract i j = #[] := by
+  intro h
+  simp only [extract, Nat.min_def, Nat.sub_eq, mkEmpty_eq]
+  split <;> rename_i h1
+  simp only [show j - i = 0 by exact Nat.sub_eq_zero_of_le h, extract_loop_0]
+  simp only [show a.size - i = 0 by
+      apply Nat.sub_eq_zero_of_le
+      apply Nat.le_trans _ h
+      simp at h1
+      exact Nat.le_of_succ_le h1,
+    extract_loop_0]
+
+theorem extract_end_le_start_size_zero {a : Array α } (h: s ≥ e) : (a.extract s e).size = 0 := by
+  rw [extract_end_le_start_eq_empty h]
+  simp
+
+theorem extract_size {a: Array α} : (e ≤ a.size) → (a.extract s e).size = e - s := by
+  intro h
+  by_cases h2: s ≤ e
+  simp [extract, Nat.min_def, h]
+  rw [Array.extract_loop_len]
+  simp?
+  rw [Nat.add_comm, Nat.sub_add_cancel]
+  exact h
+  exact h2
+  simp at h2
+  rw [extract_end_le_start_size_zero]
+  rw [Nat.sub_eq_zero_of_le]
+  exact Nat.le_of_lt h2
+  exact Nat.le_of_lt h2
+
+@[simp]
+theorem size_of_append {a b : Array α }: (a ++ b).size = a.size + b.size := by
+  simp [size]
+
+theorem get_eq_list_get {a : Array α } {i : Nat} : (h: i < a.size) → (a[i]'h = a.data[i]'h) := by
+  intro h
+  exact rfl
+
+theorem get_append_only_frist_aux {a b: Array α } (h : i < a.size) : i < (a ++ b).size := by
+  simp only [size_of_append]
+  apply Nat.lt_of_lt_of_le h
+  exact Nat.le_add_right (size a) (size b)
+
+theorem get_append_only_first {a b: Array α } (h : i < a.size):
+    (a ++ b)[i]'(get_append_only_frist_aux h) = a[i] := by
+  simp only [get_eq_list_get, append_data]
+  exact List.get_append i h
+
+/- The lemma below is still incomplete, not rearranged and fixed yet. -/
+
+theorem get_append_only_second {a b: Array α} (h : i ≥ a.size) (h2: i < a.size + b.size) :
+    (a ++ b)[i]'(sorry) = b[i - a.size]'(sorry) := by sorry
+
+theorem extract_append_only_first {a b: Array α} {i j : Nat}:
+    (j ≤ a.size) →
+    (a ++ b).extract i j = a.extract i j :=  by
+  intro h
+  by_cases h1: i ≤ j
+  let m := j - i
+  have : j = i + m := by
+    exact (Nat.sub_eq_iff_eq_add' h1).mp rfl
+  rw [this, extract_append_only_first_aux]
+  rw [← this]
+  exact h
+  simp at h1
+  rw [extract_end_le_start_eq_empty, extract_end_le_start_eq_empty]
+  apply le_of_lt h1
+  apply le_of_lt h1
+
+theorem extract_append_only_second {a b: Array α} {i j : Nat}:
+    (i ≥ a.size) →
+    (a ++ b).extract i j = b.extract (i - a.size) (j - a.size) := sorry
+
+
+
+theorem extract_mid_split {a: Array α} :
+    i ≤ j → j ≤ k → a.extract i k = a.extract i j ++ a.extract j k := by sorry
+
+theorem extract_start_overflow_eq_empty {a: Array α} (h: s ≥ a.size) :
+    a.extract s e = #[] := by sorry
+
+theorem extract_extract {a : Array α } : (s1 + s2 ≤ a.size) →
```

Comment:
This would be good to have. (Do you really need any of the inequalities? I suspect if you read off the end it still works.)